test/cipher.go

```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+		if !bytes.Equal(text, decrypted) {
+			t.Log("Encryption / Decryption failed", i)
+			t.FailNow()
+		}
+
+		bc.Message(mac, nmacs[i], nil)
+		if subtle.ConstantTimeAllEq(mac, 0) != 1 {
+			t.Log("MAC Check failed")
+			t.FailNow()
+		}
+	}
+
+	// Different keys test
+	for i := range ncrypts {
+		for j := range ncrypts {
+			if i == j {
+				continue
+			}
+			bc = newCipher(nkeys[i])
+			bc.Message(decrypted, ncrypts[j], ncrypts[j])
+			bc.Message(mac, nmacs[j], nil)
+			if subtle.ConstantTimeAllEq(mac, 0) == 1 {
+				t.Log("MAC Check passed")
+				t.FailNow()
+			}
+		}
+	}
+
+	// Not enough randomness in 1 byte to pass this consistently
+	if len(ncrypts[0]) < 8 {
+		return
+	}
+
+	// Bit difference test
+	for i := range ncrypts {
+		for j := i + 1; j < len(ncrypts); j++ {
+			res := BitDiff(ncrypts[i], ncrypts[j])
+			if res < bitdiff {
+				t.Log("Encryptions not sufficiently different", res)
+				t.FailNow()
+			}
+		}
+	}
+
+	deltacopy := make([]byte, cryptsize)
+
+	// Bit flipping test
+	for i := range ncrypts {
+		copy(ncrypts[i], deltacopy)
+
+		deltacopy[0] ^= 255
+		bc = newCipher(nkeys[i])
+		bc.Message(decrypted, deltacopy, deltacopy)
+		bc.Message(mac, nmacs[i], nil)
+		if subtle.ConstantTimeAllEq(mac, 0) == 1 {
+			t.Log("MAC Check passed")
+			t.FailNow()
+		}
+		deltacopy[0] = ncrypts[i][0]
+
+		deltacopy[len(deltacopy)/2-1] ^= 255
+		bc = newCipher(nkeys[i])
+		bc.Message(decrypted, deltacopy, deltacopy)
+		bc.Message(mac, nmacs[i], nil)
+		if subtle.ConstantTimeAllEq(mac, 0) == 1 {
+			t.Log("MAC Check passed")
+			t.FailNow()
+		}
+		deltacopy[len(deltacopy)/2-1] = ncrypts[i][len(deltacopy)/2-1]
+
+		deltacopy[len(deltacopy)-1] ^= 255
+		bc = newCipher(nkeys[i])
+		bc.Message(decrypted, deltacopy, deltacopy)
+		bc.Message(mac, nmacs[i], nil)
+		if subtle.ConstantTimeAllEq(mac, 0) == 1 {
+			t.Log("MAC Check passed")
+			t.FailNow()
+		}
+
+		deltamac := make([]byte, hashsize)
+		copy(nmacs[i], deltamac)
+		deltamac[0] ^= 255
+		bc = newCipher(nkeys[i])
+		bc.Message(decrypted, ncrypts[i], ncrypts[i])
+		bc.Message(mac, deltamac, nil)
+		if subtle.ConstantTimeAllEq(mac, 0) == 1 {
+			t.Log("MAC Check passed")
+			t.FailNow()
+		}
+	}
+}
+
+// Iterate through various sized messages and verify
+// that encryption and authentication work
+func BCAuthenticatedEncryptionHelper(t *testing.T,
+	newCipher func([]byte, ...interface{}) abstract.Cipher,
+	n int, bitdiff float64) {
+	//	AuthenticateAndEncrypt(t, newCipher, n, bitdiff, []byte{})
+	AuthenticateAndEncrypt(t, newCipher, n, bitdiff, []byte{'a'})
+	AuthenticateAndEncrypt(t, newCipher, n, bitdiff, []byte("Hello, World"))
+
+	kb := make([]byte, 2^10)
+	for i := 0; i < len(kb); i++ {
+		kb[i] = byte(i & 256)
+	}
+	AuthenticateAndEncrypt(t, newCipher, n, bitdiff, kb)
+
+	mb := make([]byte, 2^20)
+	for i := 0; i < len(mb); i++ {
+		mb[i] = byte(i & 256)
+	}
+	AuthenticateAndEncrypt(t, newCipher, n, bitdiff, mb)
+}
 
+func BlockCipherTest(t *testing.T,
```

Comment:
Why "TestBlockCipher"?  These aren't block ciphers we're testing; we're testing the "abstract.Cipher" interface, which I've been calling a "Message Cipher" interface although that may not be the best name - but in any case the name "block cipher" is certainly not appropriate here.  For now just say TestCipher probably.